### ISSUE TITLE
Option to disable no-param-reassign per function scope

### ISSUE THREAD

**Issue by dandv** [eslint/eslint#8907]
If the function body code modifies a parameter that's passed by value (Number, String, Boolean, Date), eslint should not warn that the parameter is modified, since the modifications won't survive after the function returns.

(contrived) example:
```js
function timestamp2unix(timestamp) {
  if (!timestamp) timestamp = Date.now();  // <--
  // another legitimate parameter reassignment
  if (timestamp < MIN_DATE_IN_HISTORY) timestamp = MIN_DATE_IN_HISTORY;assignment
  return new Date(timestamp).getTime();
}

**Comment by eslintbot** [comment]
Hi @dandv, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by dandv** [comment]
@platinumazure: What I would like to do is disable `no-param-reassign` for parameters passed by value. Something like

```js
'no-param-reassign': [
      'warn',
      { 'allowPassedByValue': true },
    ],
```

If ESLint can't infer the parameter passing type (e.g. from the adjacent JSDoc), then a line comment like this would solve the problem:

```js
    // eslint-disable-func no-param-reassign
```

**Comment by ilyavolodin** [comment]
@dandv ESLint can't determine if the argument was passed by value, you need type system for that. (JSDoc is unreliable, since not everyone uses it, and it can be outdated). As to inline comment, you can already do that by adding
```js
// eslint-disable-line no-param-reassign
```
next to the assignment. Or by wrapping function into `/* eslint-disable no-param-reassign */ /* eslint-enable */` comments.

**Comment by dandv** [comment]
Thanks @ilyavolodin. I meant a function-level comment like `// eslint-disable-func no-param-reassign` to disable the check for the entire function (or block) that line was found in.

Anyway, compared to your solution, my proposal would only save one line of comments, so probably the current way to wrap the function between `/* eslint-disable no-param-reassign */` and `/* eslint-enable */`  is good enough.

**Comment by not-an-aardvark** [comment]
Closing because the existing solution is satisfactory.
